qiskit/transpiler/basepasses.py

```diff
@@ -11,24 +11,92 @@
 # that they have been altered from the originals.
 
 """Base transpiler passes."""
+from __future__ import annotations
 
 from abc import abstractmethod
+from collections.abc import Callable, Hashable, Iterable
+from inspect import signature
 
-from qiskit.passmanager.base_pass import GenericPass
-from qiskit.passmanager.propertyset import PropertySet
+from qiskit.circuit import QuantumCircuit
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.passmanager.base_optimization_tasks import GenericPass, OptimizerTask
+from qiskit.passmanager.propertyset import PropertySet, PassState, RunState
 
+from .exceptions import TranspilerError
+from .fencedobjs import FencedDAGCircuit
 from .layout import TranspileLayout
 
 
-class BasePass(GenericPass):
+class MetaPass(type):
+    """Metaclass for transpiler passes.
+
+    Enforces the creation of some fields in the pass while allowing passes to
+    override ``__init__``.
+    """
+
+    # Drop this functionality in the future.
+    # This metaclass provides a pass equivalence evaluation based on the constructor arguments.
+    # This implicit fake-hash based equivalence is fragile, and the pass developer must
+    # explicitly implement equivalence check logic for each pass if necessary.
+    # Currently, this metaclass is just here for backward compatibility, because
+    # circuit pass manager has a functionality to avoid multiple execution of the
+    # same pass (even though they are scheduled so). This is managed by the valid_passes set,
+    # and executed passes are added to this collection to avoid future execution.
+    # Dropping this metaclass causes many unittest failures and thus this is
+    # considered as a breaking API change.
+    # For example, test.python.transpiler.test_pass_scheduler.TestLogPasses.test_passes_in_linear
```

Comment:
Before removing this, we must do either:
1) implementing `__eq__` method to every pass. This is straightforward but could break user provided pass which is out of our control.
2) allowing pass manager to run the same pass multiple times (I have no idea about side effect), and drop some unittests. Probably we can also drop `GenericPass.requires` that often injects pass multiple times (e.g. two passes can have a common required pass), because building proper pipeline is responsibility of user.

I would prefer 2, because implementation of proper equivalence must be difficult, i.e. all circuit passes will take `Target`, which doesn't implement equivalence check and object is mutable.

Reply:
This sounds like a sensible plan.  I suppose we could potentially emit a deprecation warning with a small amount of `__initsubclass__` trickery; if the `__eq__` method resolves to the parent class's method, we warn them, blaming the class-definition code, that they either need to implement their own `__eq__` or explicitly set `__eq__ = object.__eq__`.

We can do this in follow-ups ahead of 1.0, potentially.

Reply:
Ah that's smart. I like the idea.